File: lib/hap-parser.ts
/**
 * HAP 包解析器
 * 参考 hap_parser.py 逻辑用 TypeScript 重写
 * HAP 本质是 ZIP 压缩包
 */

import * as fs from 'fs';
import * as path from 'path';
import * as unzipper from 'unzipper';

export interface TestLib {
  arch: string;
  /** 相对于 libs/{arch}/ 的路径，如 tests/qtbase/char/libtst_qatomicinteger_char.so */
  path: string;
  name: string;
  module: string;
}

const DEFAULT_ARCHS = ['arm64-v8a', 'armeabi-v7a', 'x86_64'];

/** 从 HAP 文件中解析出所有测试库列表 */
export async function parseHap(
  hapFilePath: string,
  architectures: string[] = DEFAULT_ARCHS
): Promise<TestLib[]> {
  const testLibs: TestLib[] = [];

  const directory = await unzipper.Open.file(hapFilePath);

  for (const arch of architectures) {
    // 筛选 libs/{arch}/tests/ 下的 libtst_*.so 文件
    const prefix = `libs/${arch}/`;
    const testsPrefix = `${prefix}tests/`;

    for (const file of directory.files) {
      const entryPath = file.path;
      if (
        entryPath.startsWith(testsPrefix) &&
        entryPath.endsWith('.so')
      ) {
        const fileName = path.basename(entryPath);
        if (!fileName.startsWith('libtst_')) { continue; }

        // 相对于 libs/{arch}/ 的路径
        const relativePath = entryPath.slice(prefix.length);

        // 提取模块名：tests/{module}/...
        const parts = relativePath.split('/');
        const moduleName =
          parts.length >= 2 && parts[0] === 'tests' ? parts[1] : 'unknown';

        testLibs.push({
          arch,
          path: relativePath,
          name: fileName,
          module: moduleName,
        });
      }
    }
  }

  return testLibs;
}

/** 从已解析的列表中提取所有模块名 */
export function getModules(testLibs: TestLib[]): string[] {
  const modules = new Set<string>();
  for (const lib of testLibs) {
    if (lib.module !== 'unknown') { modules.add(lib.module); }
  }
  return Array.from(modules).sort();
}

/** 过滤测试库 */
export function filterTestLibs(
  testLibs: TestLib[],
  filterArch?: string,
  filterModule?: string | string[],
  filterPattern?: string,
  ignoreModules?: string[]
): TestLib[] {
  let result = testLibs;
  if (filterArch) { result = result.filter((t) => t.arch === filterArch); }

  const modules = Array.isArray(filterModule)
    ? filterModule.filter(Boolean)
    : filterModule ? [filterModule] : [];

  if (modules.length > 0) {
    result = result.filter((t) => modules.some((m) => t.path.startsWith(`tests/${m}/`)));
  }

  if (ignoreModules && ignoreModules.length > 0) {
    result = result.filter((t) => {
      const nameNoSo = t.name.replace(/\.so$/, '');           // libtst_qchar
      const shortName = nameNoSo.replace(/^libtst_/, '');     // qchar
      return !ignoreModules.some(
        (entry) =>
          entry === t.path ||       // tests/qtbase/tst_qpluginloader/bin/libtst_xxx.so
          entry === t.module ||
          entry === t.name ||
          entry === nameNoSo ||
          entry === shortName
      );
    });
  }

  if (filterPattern) {
    result = result.filter((t) => t.name.includes(filterPattern));
  }
  return result;
}

/**
 * 从 HAP 包内读取 resources/resfile/gitignore 文件，
 * 返回要忽略的模块名列表（过滤掉注释行和空行）。
 * 若文件不存在则返回空数组。
 */
export async function readHapIgnoreList(hapFilePath: string): Promise<string[]> {
  try {
    const directory = await unzipper.Open.file(hapFilePath);
    const entry = directory.files.find(
      (f) => f.path === 'resources/resfile/gitignore'
    );
    if (!entry) return [];
    const buf = await entry.buffer();
    const text = buf.toString('utf-8');
    return text
      .split(/\r?\n/)
      .map((line) => line.trim())
      .filter((line) => line.length > 0 && !line.startsWith('#'));
  } catch {
    return [];
  }
}

/** 保存上传的 HAP 文件到本地，返回保存路径 */
export async function saveHapFile(
  buffer: Buffer,
  fileName: string,
  uploadDir: string
): Promise<string> {
  if (!fs.existsSync(uploadDir)) {
    fs.mkdirSync(uploadDir, { recursive: true });
  }
  const savePath = path.join(uploadDir, fileName);
  fs.writeFileSync(savePath, buffer);
  return savePath;
}

File: script/hap_parser.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
HAP包解析模块
"""

import os
import zipfile
import tempfile
import shutil


class HAPParser:
    """HAP包解析器"""
    
    def __init__(self, hap_file, architectures=None):
        self.hap_file = hap_file
        self.architectures = architectures or ["arm64-v8a", "armeabi-v7a", "x86_64"]
        self.test_libs = []
    
    def extract_and_find_test_libs(self):
        """解压HAP包并查找测试库"""
        print("\n解析HAP包，查找测试库...")
        
        # 创建临时目录
        temp_dir = tempfile.mkdtemp()
        
        try:
            # 解压HAP文件
            with zipfile.ZipFile(self.hap_file, 'r') as zip_ref:
                zip_ref.extractall(temp_dir)
            
            # 查找libs目录
            libs_dir = os.path.join(temp_dir, "libs")
            if not os.path.exists(libs_dir):
                print("警告: HAP包中未找到libs目录")
                return []
            
            # 遍历架构目录
            for arch in self.architectures:
                arch_dir = os.path.join(libs_dir, arch)
                if not os.path.exists(arch_dir):
                    continue
                
                tests_dir = os.path.join(arch_dir, "tests")
                if not os.path.exists(tests_dir):
                    continue
                
                # 递归查找所有.so文件
                for root, dirs, files in os.walk(tests_dir):
                    for file in files:
                        # 只处理 libtst_ 开头的 .so 文件
                        if file.startswith('libtst_') and file.endswith('.so'):
                            # 获取相对于libs/{arch}的路径
                            full_path = os.path.join(root, file)
                            relative_path = os.path.relpath(full_path, arch_dir)
                            # 转换为Unix路径格式
                            relative_path = relative_path.replace('\\', '/')
                            
                            # 提取模块名称（从路径中获取）
                            path_parts = relative_path.split('/')
                            module = path_parts[1] if len(path_parts) >= 2 and path_parts[0] == 'tests' else 'unknown'
                            
                            self.test_libs.append({
                                'arch': arch,
                                'path': relative_path,
                                'name': file,
                                'module': module
                            })
            
            print(f"找到 {len(self.test_libs)} 个测试库")
            
            # 按架构分组显示
            for arch in self.architectures:
                arch_libs = [lib for lib in self.test_libs if lib['arch'] == arch]
                if arch_libs:
                    print(f"  {arch}: {len(arch_libs)} 个测试库")
            
            return self.test_libs
            
        finally:
            # 清理临时目录
            shutil.rmtree(temp_dir, ignore_errors=True)
    
    def get_test_libs(self):
        """获取测试库列表"""
        if not self.test_libs:
            self.extract_and_find_test_libs()
        return self.test_libs
    
    def filter_test_libs(self, filter_arch=None, filter_pattern=None, filter_module=None):
        """过滤测试库"""
        tests = self.get_test_libs()
        
        if filter_arch:
            tests = [t for t in tests if t['arch'] == filter_arch]
        
        if filter_module:
            tests = [t for t in tests if t['path'].startswith(f'tests/{filter_module}/')]
        
        if filter_pattern:
            tests = [t for t in tests if filter_pattern in t['name']]
        
        return tests
    
    def get_modules(self, test_libs=None):
        """提取所有模块名称"""
        if test_libs is None:
            test_libs = self.test_libs
        
        modules = set()
        for lib in test_libs:
            path_parts = lib['path'].split('/')
            if len(path_parts) >= 2 and path_parts[0] == 'tests':
                modules.add(path_parts[1])
        
        return sorted(modules)
